irohad/model/converters/impl/json_query_factory.cpp

```diff
@@ -45,28 +45,73 @@ namespace iroha {
           log_->error("Json is ill-formed");
           return nonstd::nullopt;
         }
-        // check if all necessary fields are there
+
         auto obj_query = doc.GetObject();
-        auto req_fields = {"signature", "creator_account_id", "created_ts",
-                           "query_counter", "query_type"};
-        if (std::any_of(req_fields.begin(), req_fields.end(),
-                        [&obj_query](auto &&field) {
-                          return not obj_query.HasMember(field);
-                        })) {
-          log_->error("No required fields in json");
-          return nonstd::nullopt;
+        // check if all necessary fields are there
+        {
+          auto req_fields = {"signature", "creator_account_id", "created_ts",
+                             "query_counter", "query_type"};
+          if (std::any_of(req_fields.begin(), req_fields.end(),
+                          [&obj_query](auto &&field) {
+                            return not obj_query.HasMember(field);
+                          })) {
+            log_->error("No required fields in json");
+            return nonstd::nullopt;
+          }
+        }
+
+        // check if all member values have valid type.
+        {
+          if (not obj_query["signature"].IsObject()) {
```

Comment:
Don't know if it will improve readability of this function, but we can use maps for this operation.
`std::unordered_map<std::string, bool(rapidjson::Document::*)()>`
And then put:
`make_pair("signature", &rapidjson::Document::IsObject)`